### DISCUSSION TITLE
GraphQL Hook Generation

### DISCUSSION THREAD

**Discussion by brentmclark** [TanStack/query#1091] [General]
This is a continuation of a conversation started in the `react-query` channel on the TANSTACK Discord server today.
[See the original discussion here](https://discordapp.com/channels/719702312431386674/719702452861009930/758682866120130603)
---
The purpose of this discussion is to talk about the implications of generating hooks using a codegen tool (probably [`graphql-code-generator`](https://graphql-code-generator.com/)) that analyzes a graphql schema. 

For starters, all of the hook generation happening for the `urql` and `apollo` plugins in `graphql-code-generator` use the `__typename` as a part of the hook name.  For example:

#### this schema

``` gql
type Person {
  id
  name
}
```

#### would become

``` javascript
const person = usePerson(id)
```

... or whatever the hook signature would be.  The important part is that the `__typename` from the schema becomes part of the name of the hook.

This works well if your cache keys are already using the `__typename` _and_ when the cache keys are static.  In RQ, however, there is intentionally no coupling between the cache key and the payload.  This is not the standard for client-side caching in GraphQL, but it _is_ the standard for react query.  People seem to enjoy this concept of decoupled cache keys; is that something that will remain desirable in a schema-generated world?

---

To summarize my long-winded question:  

**Should a GraphQL flavor of React Query continue to use a dynamic cache key, or should it lean on the type name from the schema file?**
